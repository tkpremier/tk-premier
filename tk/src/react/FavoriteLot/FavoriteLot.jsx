import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import classNames from 'classnames';
import isNull from 'lodash/isNull';
import { saveLot, deleteLot } from '../PhillipsUser/actions';
import getPhillipsBackboneProperty from '../utils/getPhillipsBackboneProperty';
import matchError from '../PhillipsUser/matchError';
import sendAnalytics from '../../utils/sendAnalytics';

const FavoriteLot = ({ lotNumberFull, saleNumber, onMouseEnter, onMouseLeave }) => {
  const bbEvent = useRef(null);
  useEffect(() => {
    getPhillipsBackboneProperty('Events')
      .then(eventEmitter => (bbEvent.current = eventEmitter))
      .catch(() => (bbEvent.current = null));
  }, []);
  const { error, favoriteLots, user } = useSelector(({ favoriteLots, user, error }) => ({
    error: matchError('FAVORITE_LOT', ({ lot }) => lot.lotNumberFull.trim() === lotNumberFull.trim())(error)
      ? error
      : null,
    favoriteLots,
    user
  }));
  const dispatch = useDispatch();
  const faveSaleIndex = favoriteLots.map(({ saleNumber: faveSaleNumber }) => faveSaleNumber).indexOf(saleNumber);
  const active = faveSaleIndex > -1 && favoriteLots[faveSaleIndex]?.lots?.indexOf(lotNumberFull.trim()) > -1;
  const toolTipMessage = isNull(error) ? (active ? 'Unfavorite lot' : 'Favorite lot') : error.message;
  const trackActivity = () => {
    const page = document ? document.title : 'placeholder';
    const analyticsData = {
      eventCategory: `Favorite Lot / ${page} / SaleNumber ${saleNumber} / LotNumber ${lotNumberFull.trim()}`,
      eventAction: 'Favorite Lot',
      eventLabel: `Email: ${user.email}`
    };
    sendAnalytics(analyticsData);
  };
  const onClick = () => {
    if (!user.loggedIn) {
      if (!isNull(bbEvent.current)) {
        bbEvent.current.trigger('openRegister');
      }
      if (typeof localStorage !== 'undefined') {
        localStorage.setItem('favoritedLot', `${saleNumber}-${lotNumberFull}`);
      }
      return;
    }
    if (active) {
      dispatch(deleteLot(user.id, { lotNumber: lotNumberFull.trim(), saleNumber }));
    } else {
      trackActivity();
      dispatch(saveLot(user.id, { lotNumber: lotNumberFull.trim(), saleNumber }));
    }
  };
  return (
    <button
      className={classNames('favorite-lot', { active })}
      data-lot-id={`${saleNumber}-${lotNumberFull}`}
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      type="button"
      tabIndex={0}
    >
      <span className={classNames('icon', { active })} />
      <span className={classNames('tooltip', { error: !isNull(error) })}>{toolTipMessage}</span>
    </button>
  );
};

FavoriteLot.defaultProps = {
  error: null,
  onMouseEnter: () => false,
  onMouseLeave: () => false
};

FavoriteLot.propTypes = {
  error: PropTypes.shape({
    errorType: PropTypes.string,
    message: PropTypes.string
  }),
  lotNumberFull: PropTypes.string.isRequired,
  user: PropTypes.shape({
    id: PropTypes.string,
    email: PropTypes.string,
    loggedIn: PropTypes.bool
  }).isRequired,
  saleNumber: PropTypes.string.isRequired,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func
};

export { FavoriteLot as UnWrappedFavoriteLot };
export default FavoriteLot;
