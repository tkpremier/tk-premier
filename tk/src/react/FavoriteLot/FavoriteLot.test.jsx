import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import toJson from 'enzyme-to-json';
import { UnWrappedFavoriteLot } from './FavoriteLot';


describe('UnWrappedFavoriteLot Button', () => {
  const mockUser = { email: 'FakeyMcFakeUser@fake.com', id: '0' };
  Enzyme.configure({ adapter: new Adapter() });
  
  it('Matches the snapshot', () => {
    const component = shallow(
      <UnWrappedFavoriteLot
        active={false}
        saleNumber="NY00000"
        lotNumberFull="1"
        department="department"
        saveLot={() => 'mock'}
        deleteLot={() => 'mock'}
      />
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('when active it should have the active class', () => {
    const component = shallow(
      <UnWrappedFavoriteLot
        active
        saleNumber="NY00000"
        lotNumberFull="1"
        department="department"
        saveLot={() => 'mock'}
        deleteLot={() => 'mock'}
      />
    );
    expect(component.hasClass('active')).toEqual(true);
  });

  it('when inactive onClick calls props.saveLot', () => {
    const saveLot = jest.fn();

    const component = shallow(
      <UnWrappedFavoriteLot
        active={false}
        saleNumber="NY00000"
        lotNumberFull="1"
        department="department"
        saveLot={saveLot}
        deleteLot={() => 'mock'}
        user={mockUser}
        userLoggedIn
      />
    );

    component.simulate('click');
    expect(saveLot).toHaveBeenCalled();
  });

  it('when active onClick calls props.deleteLot', () => {
    const deleteLot = jest.fn();

    const component = shallow(
      <UnWrappedFavoriteLot
        active
        saleNumber="NY00000"
        lotNumberFull="1"
        department="department"
        saveLot={() => 'mock'}
        deleteLot={deleteLot}
        user={mockUser}
        userLoggedIn
      />
    );

    component.simulate('click');
    expect(deleteLot).toHaveBeenCalled();
  });
  
});
