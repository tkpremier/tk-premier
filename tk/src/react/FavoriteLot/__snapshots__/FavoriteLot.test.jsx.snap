// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnWrappedFavoriteLot Button Matches the snapshot 1`] = `
<div
  className="favorite-lot"
  onClick={[Function]}
  onMouseEnter={null}
  role="button"
  tabIndex={0}
>
  <span
    className="icon"
  />
  <span
    className="tooltip"
  >
    Favorite lot
  </span>
</div>
`;
